File: 05-01-2025_Shifting_Letters_II.java
class Solution {
    public String shiftingLetters(String s, int[][] shifts) {
        int n = s.length();
        
        int[] diffArray = new int[n];
        
        for (int[] query : shifts) {
            int left = query[0];
            int right = query[1];
            int dir = query[2] == 0 ? -1 : 1;
            
            diffArray[left] += dir;
            
            if (right + 1 < n) {
                diffArray[right + 1] -= dir;
            }
        }
        
        int prefixSum = 0;
        StringBuilder result = new StringBuilder(s);
        
        for (int i = 0; i < n; i++) {
            int pos = result.charAt(i) - 'a';
            
            prefixSum += diffArray[i];
            int newPos = (pos + prefixSum) % 26;
            
            if (newPos < 0) {
                newPos += 26;
            }
            
            result.setCharAt(i, (char) ('a' + newPos));
        }
        
        return result.toString();
        
    }
}

File: 05-01-2025-Shifting_Letters_II.cpp
class Solution {
public:
    string shiftingLetters(string s, vector<vector<int>>& shifts) 

    {
        int n = s.size();

        vector <int> diffarray(n,0);

        for(auto query : shifts)
        {
            int left = query[0];

            int right = query[1];
            int dir = 1;
            if(query[2]==0)
            {
                dir = -1;
            }

            diffarray[left]+=dir;

            if(right+1<n)
            {
                diffarray[right+1] -=dir;
            }
        } 

        int prefixsum = 0;

         for(int i=0;i<n;i++)
         {  
            int pos = s[i]-'a';

            prefixsum+=diffarray[i];

            int newpos = (pos+prefixsum)%26;

            if(newpos<0)
            {
                newpos+=26;
            }
            
            s[i] = 'a'+newpos;


            
         }

         return s;
    }
};
